src/librustc/mir/interpret/allocation.rs

```diff
@@ -324,6 +324,33 @@ impl<'tcx, Tag: Copy, Extra: AllocationExtra<Tag>> Allocation<Tag, Extra> {
         })
     }
 
+    /// Reads bytes until a `0x00` is encountered. Will error if the end of the allocation
+    /// is reached before a `0x00` is found.
+    ///
+    /// Most likely, you want to call `Memory::read_wide_str` instead of this method.
+    pub fn read_wide_str(
+        &self,
+        cx: &impl HasDataLayout,
+        ptr : Pointer<Tag>
+    ) -> InterpResult<'tcx, &[u8]>
+    {
+        assert_eq!(ptr.offset.bytes() as usize as u64, ptr.offset.bytes());
+        let offset = ptr.offset.bytes() as usize;
+        Ok(match self.bytes[offset..self.bytes.len() - 1].iter().step_by(2)
+           .zip(self.bytes[(offset+1)..].iter().step_by(2))
```

Comment:
That's a clever iterator construction! Please add a comment explaining what is going on (that you are setting up an iterator yielding pairs of adjacent bytes).

Is the `self.bytes.len() - -1` on the first iterator really needed? The *shorter* of the two iterator will stop zipping, so I think you don't need that, but I might be wrong.

Reply:
Thanks for your review! I'll add a comment to explain the iterator part.
As you mentioned, `self.bytes.len() - 1` is not necessary here. I will remove that part soon.